source/extensions/filters/http/cache/http_cache.h

```diff
@@ -0,0 +1,322 @@
+#pragma once
+
+#include <iosfwd>
+#include <string>
+#include <vector>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/time.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+
+#include "common/common/assert.h"
+
+#include "source/extensions/filters/http/cache/key.pb.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+// Whether a given cache entry is good for the current request.
+enum class CacheEntryStatus {
+  // This entry is fresh, and an appropriate response to the request.
+  Ok,
+  // No usable entry was found. If this was generated for a cache entry, the
+  // cache should delete that entry.
+  Unusable,
+  // This entry is stale, but appropriate for validating
+  RequiresValidation,
+  // This entry is fresh, and an appropriate basis for a 304 Not Modified
+  // response.
+  FoundNotModified,
+  // This entry is fresh, but can't satisfy the requested range(s).
+  UnsatisfiableRange,
+};
+
+std::ostream& operator<<(std::ostream& os, CacheEntryStatus status);
+
+// Byte range from an HTTP request.
+class RawByteRange {
+public:
+  // - If first==UINT64_MAX, construct a RawByteRange requesting the final last body bytes.
+  // - Otherwise, construct a RawByteRange requesting the [first,last] body bytes.
+  // Prereq: first == UINT64_MAX || first <= last
+  // Invariant: isSuffix() || firstBytePos() <= lastBytePos
+  // Examples: RawByteRange(0,4) requests the first 5 bytes.
+  //           RawByteRange(UINT64_MAX,4) requests the last 4 bytes.
+  RawByteRange(uint64_t first, uint64_t last) : first_byte_pos_(first), last_byte_pos_(last) {
+    RELEASE_ASSERT(isSuffix() || first <= last, "Illegal byte range.");
```

Comment:
nit: normal ASSERT here and in similar cases. We should use normal ASSERT for programming errors. Similar comment for below and elsewhere.

Reply:
Done.